governance/program/src/processor/process_add_signatory.rs

```diff
@@ -25,26 +30,49 @@ pub fn process_add_signatory(
     let account_info_iter = &mut accounts.iter();
 
     let proposal_info = next_account_info(account_info_iter)?; // 0
-    let token_owner_record_info = next_account_info(account_info_iter)?; // 1
-    let governance_authority_info = next_account_info(account_info_iter)?; // 2
+    let signatory_record_info = next_account_info(account_info_iter)?; // 1
 
-    let signatory_record_info = next_account_info(account_info_iter)?; // 3
+    let governance_info = next_account_info(account_info_iter)?; // 2
 
-    let payer_info = next_account_info(account_info_iter)?; // 4
-    let system_info = next_account_info(account_info_iter)?; // 5
+    let payer_info = next_account_info(account_info_iter)?; // 3
+    let system_info = next_account_info(account_info_iter)?; // 4
 
-    let rent = Rent::get()?;
+    let governance_data = get_governance_data(program_id, governance_info)?;
 
-    let mut proposal_data = get_proposal_data(program_id, proposal_info)?;
+    let mut proposal_data =
+        get_proposal_data_for_governance(program_id, proposal_info, governance_info.key)?;
     proposal_data.assert_can_edit_signatories()?;
 
-    let token_owner_record_data = get_token_owner_record_data_for_proposal_owner(
-        program_id,
-        token_owner_record_info,
-        &proposal_data.token_owner_record,
-    )?;
+    if !signatory_record_info.data_is_empty() {
+        return Err(GovernanceError::SignatoryRecordAlreadyExists.into());
+    }
+
+    // All required signatories must be added before additional signatories can be added
+    if proposal_data.signatories_count < governance_data.required_signatories_count {
```

Comment:
You could circumvent required signatories by 1) Adding all required signatories first 2) then optional ones and 3) then removing the required ones.
Probably the best solution would be to remove the `RemoveSignatory` instruction.  It doesn't add much value and creates a potential security risk

Reply:
Yeah I agree

Reply:
Removed it.